clang/lib/Analysis/UnsafeBufferUsage.cpp

```diff
@@ -927,21 +927,28 @@ class CArrayToPtrAssignmentGadget : public FixableGadget {
 /// over one of its pointer parameters.
 class UnsafeBufferUsageAttrGadget : public WarningGadget {
   constexpr static const char *const OpTag = "call_expr";
-  const CallExpr *Op;
+  const Expr *Op;
 
 public:
   UnsafeBufferUsageAttrGadget(const MatchFinder::MatchResult &Result)
       : WarningGadget(Kind::UnsafeBufferUsageAttr),
-        Op(Result.Nodes.getNodeAs<CallExpr>(OpTag)) {}
+        Op(Result.Nodes.getNodeAs<Expr>(OpTag)) {}
 
   static bool classof(const Gadget *G) {
     return G->getKind() == Kind::UnsafeBufferUsageAttr;
   }
 
   static Matcher matcher() {
+    auto HasUnsafeFielDecl = 
+        member(fieldDecl(allOf(
+               anyOf(hasPointerType(), hasArrayType()),
```

Comment:
Yeah for now we don't respond to the attribute on member variables that aren't raw pointers or raw arrays. I think eventually we should allow it on size integer fields too, because you can get an OOB access by setting the size to the incorrect value, if that value is actually used for runtime checks (which may or may not be the libc++ hardening checks). However, this is trickier because many uses of such fields are actually benign. In theory I'm not opposed to simply allowing it on arbitrary fields, but again, it's hard to determine which ones are benign.

There are also some benign cases when it comes to raw pointers and arrays too. Eg., comparing a raw pointer to null shouldn't trigger a warning even if the raw pointer is annotated as unsafe to use. It's simply never a bad thing to do. So I expect this code to get more complex in the future. And we probably need to see if we run into a lot of benign cases before we roll this out(?)

Now, the real question is, do we need to emit a warning (or even an error) when the attribute isn't placed on a field of a supported type? If we plan to emit such warnings in the future, we really need to do that right now, because it's considered to be a major annoyance when you suddenly ban previously correct code. So I think we gotta do this as part of the initial patch.

Reply:
Doesn't the absence of `-Wunsafe-buffer-usage` warning give more or less the same signal as the additional warning would provide?
I suggest we don't implement the extra warning.

Reply:
I was discussing offline about this with Devin and he said it is a good idea to not restrict what types of fields we allow the attribute to be attached to. This could be useful if someone accidentally only updates a field that stores the bounds information.

Reply:
Also, we do not emit any warning when it is placed on other types, which could be confusing. I will handle this in the next version and allow it to be added to all types.